src/smoothscroll.js

```diff
@@ -1,371 +1,436 @@
+    scroll: w.scroll || w.scrollTo,
+    scrollBy: w.scrollBy,
+    elementScroll: Element.prototype.scroll || scrollElement,
+    scrollIntoView: Element.prototype.scrollIntoView
+  };
+
+  // define timing method
+  var now = w.performance && w.performance.now
+    ? w.performance.now.bind(w.performance)
+    : Date.now;
+
+  /**
+   * changes scroll position inside an element
+   * @method scrollElement
+   * @param {Number} x
+   * @param {Number} y
+   * @returns {undefined}
    */
+  function scrollElement(x, y) {
+    this.scrollLeft = x;
+    this.scrollTop = y;
+  }
 
-  // polyfill
-  function polyfill() {
-    // return when scrollBehavior interface is supported
-    if ('scrollBehavior' in d.documentElement.style) {
-      return;
+  /**
+   * returns result of applying ease math function to a number
+   * @method ease
+   * @param {Number} k
+   * @returns {Number}
+   */
+  function ease(k) {
+    return 0.5 * (1 - Math.cos(Math.PI * k));
+  }
+
+  /**
+   * indicates if a smooth behavior should be applied
+   * @method shouldBailOut
+   * @param {Number|Object} firstArg
+   * @returns {Boolean}
+   */
+  function shouldBailOut(firstArg) {
+    if (firstArg === null
+      || typeof firstArg !== 'object'
+      || firstArg.behavior === undefined
+      || firstArg.behavior === 'auto'
+      || firstArg.behavior === 'instant') {
+      // first argument is not an object/null
+      // or behavior is auto, instant or undefined
+      return true;
     }
 
-    /**
-     * indicates if a the current browser is made by Microsoft
-     * @method isMicrosoftBrowser
-     * @param {string} userAgent
-     * @returns {Boolean}
-     */
-    function isMicrosoftBrowser(userAgent) {
-      var userAgentPatterns = ['MSIE ', 'Trident/', 'Edge/'];
-      return new RegExp(userAgentPatterns.join('|')).test(userAgent);
+    if (typeof firstArg === 'object' && firstArg.behavior === 'smooth') {
+      // first argument is an object and behavior is smooth
+      return false;
     }
 
-    /*
-     * globals
-     */
-    var Element = w.HTMLElement || w.Element;
-    var SCROLL_TIME = 468;
-    var ROUNDING_TOLERANCE = isMicrosoftBrowser(w.navigator.userAgent) ? 1 : 0;
-
-    /*
-     * object gathering original scroll methods
-     */
-    var original = {
-      scroll: w.scroll || w.scrollTo,
-      scrollBy: w.scrollBy,
-      elScroll: Element.prototype.scroll || scrollElement,
-      scrollIntoView: Element.prototype.scrollIntoView
-    };
-
-    /*
-     * define timing method
-     */
-    var now = w.performance && w.performance.now
-      ? w.performance.now.bind(w.performance) : Date.now;
-
-    /**
-     * changes scroll position inside an element
-     * @method scrollElement
-     * @param {Number} x
-     * @param {Number} y
-     */
-    function scrollElement(x, y) {
-      this.scrollLeft = x;
-      this.scrollTop = y;
+    // throw error when behavior is not supported
+    throw new TypeError('behavior not valid');
```

Comment:
Here are what Chrome and Firefox through when given an invalid behavior value of `'lol'`. Can we modify our message in this direction? The `Firefox` message seems *easier* since we don’t need to keep a handle on what element type or method was called.

Firefox TypeError
```
'behavior' member of ScrollOptions 'lol' is not a valid value for enumeration ScrollBehavior.
```

Chrome TypeError
```
Failed to execute 'scrollBy' on 'Element': The provided value 'lol' is not a valid enum value of type ScrollBehavior.
```